src/xform_to_datalog/datalog_facts.cc

```diff
@@ -0,0 +1,149 @@
+// #TODO(#107): In the interest of prototyping speed, I wrote out the
+//  entirety of the traversal of instantiation structures as one big
+//  traversal. This works, but it is hard to read and hard to test. We should
+//  break this up.
+DatalogFacts DatalogFacts::CreateFromManifestProto(
+    const arcs::ManifestProto &manifest_proto) {
+  std::vector<ir::TagClaim> result_claims;
+  std::vector<ir::TagCheck> result_checks;
+  std::vector<ir::Edge> result_edges;
+  absl::flat_hash_map<std::string, amt::ParticleSpec> particle_specs;
+  for (const arcs::ParticleSpecProto &particle_spec_proto :
+    manifest_proto.particle_specs()) {
+    amt::ParticleSpec particle_spec =
+        amt::ParticleSpec::CreateFromProto(particle_spec_proto);
+    std::string particle_spec_name = particle_spec.name();
+    auto ins_res = particle_specs.insert({
+      std::move(particle_spec_name), std::move (particle_spec) });
+    CHECK(ins_res.second) << "Name collision on particle spec.";
+  }
+
+  const std::string generated_recipe_name_prefix = "GENERATED_RECIPE_NAME";
+  uint64_t recipe_num = 0;
+  for (const arcs::RecipeProto &recipe_proto : manifest_proto.recipes()) {
+    const std::string &recipe_name =
+        recipe_proto.name().empty()
+          ? absl::StrCat(generated_recipe_name_prefix, recipe_num++)
+          : recipe_proto.name();
+
+    uint64_t particle_num = 0;
+    for (const arcs::ParticleProto &particle_proto : recipe_proto.particles()) {
+      const std::string &particle_spec_name = particle_proto.spec_name();
+      CHECK(!particle_spec_name.empty())
+        << "Particle with empty spec_name field not allowed.";
+      const std::string particle_name =
+          absl::StrCat(particle_spec_name, "#", particle_num++);
+
+      absl::flat_hash_map<ir::AccessPathRoot, ir::AccessPathRoot>
+          instantiation_map;
+      for (const arcs::HandleConnectionProto &connection_proto :
+        particle_proto.connections()) {
+        std::string handle_spec_name = connection_proto.name();
+        CHECK(!handle_spec_name.empty())
+          << "Handle connection with empty name field not allowed.";
+        std::string handle_name = connection_proto.handle();
+        CHECK(!handle_name.empty())
+          << "Handle with empty handle field not allowed.";
+
+        raksha::ir::HandleConnectionSpecAccessPathRoot spec_handle_root(
+            particle_spec_name, handle_spec_name);
+        raksha::ir::HandleConnectionAccessPathRoot
+          instantiated_handle_root(
+              recipe_name, particle_name, std::move(handle_spec_name));
+        raksha::ir::HandleAccessPathRoot handle_root(
+            recipe_name, std::move(handle_name));
+        CHECK(connection_proto.has_type())
+          << "Handle connection with absent type not allowed.";
+        std::unique_ptr<types::Type> connection_type =
+            types::Type::CreateFromProto(connection_proto.type());
+        ir::AccessPathSelectorsSet access_path_selectors_set =
```

Comment:
`const &`?

Reply:
Not necessary; `GetAccessPathSelectorsSet` always returns a copy-elided `AccessPathSelectorsSet`. That means that we can just use the value instead of a const reference. That is helpful because we don't need to worry about the lifetime of the set.